index.js

```diff
@@ -254,5 +200,76 @@ module.exports = {
 
   includedCommands: function() {
     return require('./lib/commands');
-  }
+  },
+
+  included: function(app) {
+    this.calculateConfig(app);
+  },
+
+  // Configuration is only available by public API in `app` passed to `included` hook.
+  // We calculate configuration in `included` hook and use it in `serverMiddleware`
+  // and `contentFor` hooks, which are executed later. This is necessary cause Ember CLI
+  // does not provide a public API to read build time configuation (`ember-cli-build.js`)
+  // yet. `this._findHost(this).options` seems to be the only reliable way to get it in
+  // these hooks but is private API.
+  calculateConfig(app) {
```

Comment:
Can we make this a "pure function", that we could then import from the test and call directly? It seems like we could test a ton of permutations that way, and keep the test coupling to implementation a bit smaller...

```js
// ...snip...
included(app) {
  this._config = calculateConfig(app.env, app.options || {}, this.project.config(app.env));
},

// ...snip...

// below main `module.exports = ` line

function calculateConfig(environment, buildConfig, runConfig, ui) {
  // current implementation
  return config;
}
module.exports._calculateConfig = calculateConfig;
```

Then if you do this, you don't need an "app mock", you can instead pass the various permutations of POJO's directly into the `calculateConfig` method.